pycds/alembic/versions/7a3b247c577b_add_varsperhistory_native_matview.py

```diff
@@ -0,0 +1,71 @@
+"""Add VarsPerHistory native matview
+
+Revision ID: 7a3b247c577b
+Revises: 8fd8f556c548
+Create Date: 2021-01-18 17:06:14.064487
+
+"""
+import logging
+from alembic import op
+import sqlalchemy as sa
+from pycds import get_schema_name
+from pycds.materialized_views import VarsPerHistory
+from pycds.alembic.helpers import db_supports_matviews
+
+# revision identifiers, used by Alembic.
+revision = "7a3b247c577b"
+down_revision = "8fd8f556c548"
+branch_labels = None
+depends_on = None
+
+logger = logging.getLogger("alembic")
+
+schema_name = get_schema_name()
+
+
+def upgrade():
+    engine = op.get_bind().engine
+    if db_supports_matviews(engine):
+        logger.debug("This database supports matviews")
+        op.drop_table("vars_per_history_mv", schema=schema_name)
+        op.create_native_materialized_view(VarsPerHistory, schema=schema_name)
+        for index in VarsPerHistory.__table__.indexes:
+            op.create_index(
+                index_name=index.name,
+                table_name=index.table.name,
+                columns=[col.name for col in index.columns],
+                unique=index.unique,
+                schema=schema_name,
+            )
+    else:
+        logger.info("This database does not support matviews: skipping upgrade")
+
+
+def downgrade():
+    engine = op.get_bind().engine
+    if db_supports_matviews(engine):
+        logger.debug("This database supports matviews")
+        for index in VarsPerHistory.__table__.indexes:
+            op.drop_index(
+                index_name=index.name,
+                table_name=index.table.name,
+                schema=schema_name,
+            )
+        op.drop_native_materialized_view(VarsPerHistory, schema=schema_name)
+        op.create_table(
```

Comment:
The downgrade does have the table. It deletes the matview and replaces it with a table.

Reply:
But only inside the `if db_supports_matviews(engine):` block. I'm arguing that the migration should ensure that table is created always. Whether the database supports matviews or not.

Reply:
Terminology: A *migration* is a sequence of revisions. A *revision* is like a git commit: it builds on previous revisions. This revision is concerned with the matview. A previous revision created the table. So all this revision requires is to drop the previously created table and replace it with the matview. The downgrade does the reverse.

When any database (including this one) is at a late stage placed under Alembic management, we have to "stamp" it with the appropriate revision number (reflecting its current state, achieved ad-hoc) so that Alembic knows where things stand and where to start applying new revisions. (There's an Alembic `stamp` command.) Ironically, that is almost certainly this very revision, since the matview it's responsible for has already been created ad-hoc.